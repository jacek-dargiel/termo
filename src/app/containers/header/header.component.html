<header>
  <button
    class="refresh"
  >
    <div
      class="refresh__progress"
      [style.width.%]="progress | async"
    ></div>
    <span class="refresh__label">Odśwież</span>
  </button>
</header>