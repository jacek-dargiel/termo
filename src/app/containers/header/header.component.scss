@import '../../../styles/variables';

:host {
  position: absolute;
  top: 0;
  display: block;
  width: 100%;
  z-index: 2;
}

header {
  padding: 0 1.5rem;
  height: $headerHeight;
  // background: $colorPrimary;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: flex-end;

  @media (max-width: 1024px) {
    padding: 0 1rem;
  }
}

.refresh {
  background: lighten($colorPrimary, 10);
  box-shadow: 0 0 2rem 0.5rem rgba(darken($colorPrimary, 30), 0.5);
  border: 0;
  color: #fff;
  text-transform: uppercase;
  font-weight: 500;
  font-size: 1.2rem;
  padding: 0.75rem 4rem;
  position: relative;

  &:focus,
  &:active {
    outline: 0.2rem solid #eede25;
  }

  &__progress {
    position: absolute;
    width: 0;
    background-color: lighten($colorPrimary, 20);
    height: 100%;
    top: 0;
    left: 0;
    z-index: 1;
    transition: width 1s linear;

    &--refreshing {
      background-color: lighten($colorPrimary, 10);
      background-image: linear-gradient(-45deg, rgba(255, 255, 255, 0.35) 25%, rgba(255, 255, 255, 0) 25%, rgba(255, 255, 255, 0) 50%, rgba(255, 255, 255, 0.35) 50%, rgba(255, 255, 255, 0.35) 75%, rgba(255, 255, 255, 0) 75%, rgba(255, 255, 255, 0));
      background-size: 20px 20px;
      animation: candystripe 1s linear infinite;
      width: 100% !important;
      transition: width 0s linear;
    }
  }
  

  &__label {
    z-index: 2;
    position: relative;
  }
}

@keyframes candystripe {
  to {background-position: 20px 0;}
}
