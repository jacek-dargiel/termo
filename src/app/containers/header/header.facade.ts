import { Injectable } from '@angular/core';
import { RefreshSignalService } from '../../services/refresh-signal.service';
import { map, first } from 'rxjs/operators';
import { environment } from 'environments/environment';
import { Store } from '@ngrx/store';
import { State } from '../../state/reducers';
import { selectMeasurmentsLoading, selectAllLocations } from '../../state/selectors';
import { RefreshMeasurmentsStart } from '../../state/location/location.actions';

@Injectable({
  providedIn: 'root',
})
export class HeaderFacade {
  public progress = this.refreshSignal.signal.pipe(
    map(timeLeft => timeLeft * 1000),
    map(timeLeft => (environment.refreshTimeout - timeLeft) / environment.refreshTimeout),
  );
  public refreshing = this.store.select(selectMeasurmentsLoading);

  constructor (
    private refreshSignal: RefreshSignalService,
    private store: Store<State>
  ) {}

  refresh() {
    this.store.select(selectAllLocations).pipe(
      first(),
    )
      .subscribe(locations => this.store.dispatch(new RefreshMeasurmentsStart({ locations })));
  }
}
