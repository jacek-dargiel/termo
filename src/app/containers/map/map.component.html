<termo-header></termo-header>
<termo-map-location
  *ngFor="let location of locations$ | async"
  [location]="location"
  [loading]="loading$ | async"
  [selected]="(selectedLocation$ | async)?.id === location.id"
  (select)="onLocationSelect(location)"
></termo-map-location>
