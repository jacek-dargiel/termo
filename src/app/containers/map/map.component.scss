@import '../../../styles/variables';

:root {
    --mapBackgroundRatio: 9/16;
}

:host {
    position: relative;
    display: block;
    background: url($mapBackground);
    background-size: cover;
    margin-top: 1vw;
    width: 98vw;
    height: calc(98vw * var(--mapBackgroundRatio));
    box-shadow: 0 0 2rem 0.5rem rgba(darken($mapBackgroundColor, 80), 0.35);

    @media (max-width: $breakpointMobile) {
        height: auto;
    }
}
