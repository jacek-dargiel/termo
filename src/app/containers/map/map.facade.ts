import { Injectable } from '@angular/core';
import { Store } from '@ngrx/store';
import { isEmpty } from '../../helpers/lodash';
import { State } from '../../state/reducers';
import { MapInitialized, SelectLocation } from '../../state/location/location.actions';
import * as selectors from '../../state/selectors';
import { MapBackgroundService } from '../../services/map-background.service';
import { environment } from 'environments/environment';
import { Location } from '../../state/location/location.model';
import { ErrorHandlingService } from '../../services/error-handling.service';
import { tap, first } from 'rxjs/operators';

@Injectable()
export class MapFacade {
  public loading$ = this.store.select(selectors.selectLocationLoading);
  public locations$ = this.store.select(selectors.selectLocationsMappedWithKeyMeasurmentValues);
  public selectedLocation$ = this.store.select(selectors.selectSelectedLocation);

  private measurmentsByLocation$ = this.store.select(selectors.selectMeasurmentsByLocation);

  constructor(
    private store: Store<State>,
    private mapBackgroundService: MapBackgroundService,
    private errorHandlingService: ErrorHandlingService,
  ) {}

  dispatchMapInit() {
    this.store.dispatch(new MapInitialized());
  }

  getImageDimentions() {
    let url = environment.mapBackgroundUrl;
    return this.mapBackgroundService.getImageDimentions(url);
  }

  selectLocation(location: Location) {
    return this.measurmentsByLocation$.pipe(
      first(),
      tap(measurmentsByLocation => {
        let locationMeasurments = measurmentsByLocation[location.id];
        if (isEmpty(locationMeasurments)) {
          this.errorHandlingService.handle(new Error('Brak aktualnych danych do wy≈õwietlenia na wykresie.'));
          return;
        }
        this.store.dispatch(new SelectLocation({location}));
      })
    );
  }
}
