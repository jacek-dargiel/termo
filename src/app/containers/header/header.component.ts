import { Component } from '@angular/core';
import { HeaderFacade } from './header.facade';

@Component({
    selector: 'termo-header',
    templateUrl: './header.component.html',
    styleUrls: ['./header.component.scss'],
    standalone: false
})
export class HeaderComponent {
  public progress$ = this.headerFacade.progress;
  public refreshing$ = this.headerFacade.refreshing;

  constructor(
    private headerFacade: HeaderFacade,
  ) { }

  public refresh () {
    this.headerFacade.refresh();
  }

}
