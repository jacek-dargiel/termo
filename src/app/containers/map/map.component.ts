import { Component, OnInit } from '@angular/core';
import { MapFacade } from './map.facade';

@Component({
  selector: 'termo-map',
  templateUrl: './map.component.html',
  styleUrls: ['./map.component.scss']
})
export class MapComponent implements OnInit {

  constructor(
    private mapFacade: MapFacade,
  ) { }

  ngOnInit() {
    this.mapFacade.fetchLocations();
  }

}
