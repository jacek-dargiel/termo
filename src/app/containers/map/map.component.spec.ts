import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { Component, Input, HostBinding, Output, EventEmitter, CUSTOM_ELEMENTS_SCHEMA, input } from '@angular/core';

import { MapComponent } from './map.component';
import { MapFacade } from './map.facade';
import { of, Subject, BehaviorSubject } from 'rxjs';
import { LocationWithKeyMeasurmentValues, Location } from '../../state/location/location.model';
import { By } from '@angular/platform-browser';
import { HeaderComponent } from '../header/header.component';
import { TERMO_CURRENT_TIME_FACTORY } from '../../pipes/current-time.injection-token';
import { MapLocationComponent } from '../../components/map-location/map-location.component';

class MockMapFacade {
  public locationsLoading$ = new BehaviorSubject<string[]>([]);
  public locations$ = new BehaviorSubject<LocationWithKeyMeasurmentValues[]>([]);
  public selectedLocation$ = new Subject<Location>();

  public dispatchMapInit = jest.fn();
  public selectLocation = jest.fn(() => of());

  public getImageDimentions = jest.fn(() => {
    return of({width: 800, height: 600});
  });
}

let mockLocations: LocationWithKeyMeasurmentValues[] = [
  {
    id: 'temperatura.gorny_tunel',
    name: 'GÃ³rny Tunel',
    mapPosition: {
      x: 0.25,
      y: 0.75,
    },
    updatedAt: new Date('2018-09-19T22:10:00'),
    lastMeasurmentValue: 21.12,
    minimalMeasurmentValue: 19.5,
  },
  {
    id: 'temperatura.dolny_tunel',
    name: 'Dolny Tunel',
    mapPosition: {
      x: 0.5,
      y: 0.5,
    },
    updatedAt: new Date('2018-09-19T22:10:00'),
    lastMeasurmentValue: 15,
    minimalMeasurmentValue: -5,
  },
];

@Component({
    selector: 'termo-map-location',
    template: ``,
})
export class MockMapLocationComponent {
  readonly location = input<LocationWithKeyMeasurmentValues>();
  @Input() loading: boolean;
  @Input()
  @HostBinding('class.location--selected')
  selected: boolean;
  @Output() selectLocation = new EventEmitter<Location>();
}


describe('MapComponent', () => {
  let component: MapComponent;
  let fixture: ComponentFixture<MapComponent>;
  let facade: MockMapFacade;
  let since = new Date('2018-09-19T22:15:00');

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      imports: [
        MapComponent,
      ],
      providers: [
        { provide: MapFacade, useClass: MockMapFacade },
        { provide: TERMO_CURRENT_TIME_FACTORY, useValue: () => since }
      ],
      schemas: [CUSTOM_ELEMENTS_SCHEMA]
    })
    .overrideComponent(MapComponent, { remove: { imports: [HeaderComponent]}})
    .overrideComponent(MapComponent, { remove: { imports: [MapLocationComponent] }, add: { imports: [MockMapLocationComponent]}})
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(MapComponent);
    component = fixture.componentInstance;
    component.updateMapRatio = jest.fn(component.updateMapRatio);
    facade = TestBed.get(MapFacade);
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should properly render initial state', () => {
    expect(fixture).toMatchSnapshot();
  });
  it('should render locations', () => {
    facade.locations$.next(mockLocations);
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
  it('should render loading states', () => {
    facade.locations$.next(mockLocations);
    facade.locationsLoading$.next([
      'temperatura.gorny_tunel',
      'temperatura.dolny_tunel',
    ]);
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();

    facade.locationsLoading$.next([
      'temperatura.gorny_tunel',
    ]);
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();

    facade.locationsLoading$.next([
      'temperatura.dolny_tunel',
    ]);
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();

    facade.locationsLoading$.next([]);
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch map initialization action on init', () => {
    expect(facade.dispatchMapInit).toHaveBeenCalled();
  });

  it('should update map image dimentions on init', () => {
    expect(facade.getImageDimentions).toHaveBeenCalled();
    expect(component.updateMapRatio).toHaveBeenCalledWith({width: 800, height: 600});
  });

  it('should select location when clicked', () => {
    facade.locations$.next(mockLocations);
    fixture.detectChanges();
    let locationComp: MockMapLocationComponent = fixture.debugElement.query(By.css('termo-map-location')).componentInstance;
    locationComp.selectLocation.emit();
    fixture.detectChanges();
    expect(facade.selectLocation).toHaveBeenCalledWith(locationComp.location());
  });
});
