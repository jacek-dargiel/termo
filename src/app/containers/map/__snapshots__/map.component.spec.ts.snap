// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`MapComponent should properly render initial state 1`] = `
<termo-map
  el={[Function ElementRef]}
  loading$="undefined"
  locations$={[Function BehaviorSubject]}
  mapFacade={[Function MockMapFacade]}
  selectedLocation$={[Function Subject]}
  updateMapRatio={[Function Function]}
>
  <termo-header />
</termo-map>
`;

exports[`MapComponent should render loading states 1`] = `
<termo-map
  el={[Function ElementRef]}
  loading$="undefined"
  locations$={[Function BehaviorSubject]}
  mapFacade={[Function MockMapFacade]}
  selectedLocation$={[Function Subject]}
  updateMapRatio={[Function Function]}
>
  <termo-header /><termo-map-location /><termo-map-location />
</termo-map>
`;

exports[`MapComponent should render loading states 2`] = `
<termo-map
  el={[Function ElementRef]}
  loading$="undefined"
  locations$={[Function BehaviorSubject]}
  mapFacade={[Function MockMapFacade]}
  selectedLocation$={[Function Subject]}
  updateMapRatio={[Function Function]}
>
  <termo-header /><termo-map-location /><termo-map-location />
</termo-map>
`;

exports[`MapComponent should render loading states 3`] = `
<termo-map
  el={[Function ElementRef]}
  loading$="undefined"
  locations$={[Function BehaviorSubject]}
  mapFacade={[Function MockMapFacade]}
  selectedLocation$={[Function Subject]}
  updateMapRatio={[Function Function]}
>
  <termo-header /><termo-map-location /><termo-map-location />
</termo-map>
`;

exports[`MapComponent should render loading states 4`] = `
<termo-map
  el={[Function ElementRef]}
  loading$="undefined"
  locations$={[Function BehaviorSubject]}
  mapFacade={[Function MockMapFacade]}
  selectedLocation$={[Function Subject]}
  updateMapRatio={[Function Function]}
>
  <termo-header /><termo-map-location /><termo-map-location />
</termo-map>
`;

exports[`MapComponent should render locations 1`] = `
<termo-map
  el={[Function ElementRef]}
  loading$="undefined"
  locations$={[Function BehaviorSubject]}
  mapFacade={[Function MockMapFacade]}
  selectedLocation$={[Function Subject]}
  updateMapRatio={[Function Function]}
>
  <termo-header /><termo-map-location /><termo-map-location />
</termo-map>
`;
